File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/memory/ingest/autolink/TagStrategy.kt
/** Tag overlap linking — finds entries sharing >= N tags. KSA-190. */
package com.codeintel.memory.ingest.autolink

import java.sql.Connection

class TagStrategy(private val conn: Connection) : LinkingStrategy {

    override val name = "tag"

    override fun isEnabled(config: AutoLinkConfig): Boolean =
        config.tag.enabled

    override fun findCandidates(entryId: Long, config: AutoLinkConfig): List<CandidateEdge> {
        val myTags = getEntryTags(entryId)
        if (myTags.size < config.tag.minOverlap) return emptyList()

        val rows = findEntriesWithOverlappingTags(entryId, myTags)
        return rows
            .mapNotNull { (otherId, otherTags) ->
                buildTagEdge(myTags, otherTags, otherId, config)
            }
            .sortedByDescending { it.score }
            .take(config.tag.maxEdges)
    }

    private fun getEntryTags(entryId: Long): List<String> {
        val ps = conn.prepareStatement(
            "SELECT tags FROM knowledge_entries WHERE id = ?"
        )
        ps.setLong(1, entryId)
        val rs = ps.executeQuery()
        val tags = if (rs.next()) parseTags(rs.getString("tags")) else emptyList()
        rs.close(); ps.close()
        return tags
    }

    private fun findEntriesWithOverlappingTags(
        entryId: Long,
        myTags: List<String>
    ): List<Pair<Long, List<String>>> {
        val conditions = myTags.joinToString(" OR ") { "tags LIKE ?" }
        val sql = """
            SELECT id, tags FROM knowledge_entries
            WHERE id != ? AND archived_at IS NULL AND ($conditions)
        """.trimIndent()
        val ps = conn.prepareStatement(sql)
        ps.setLong(1, entryId)
        myTags.forEachIndexed { i, tag -> ps.setString(i + 2, "%$tag%") }
        val rs = ps.executeQuery()
        val results = mutableListOf<Pair<Long, List<String>>>()
        while (rs.next()) {
            results.add(rs.getLong("id") to parseTags(rs.getString("tags")))
        }
        rs.close(); ps.close()
        return results
    }

    private fun buildTagEdge(
        myTags: List<String>,
        otherTags: List<String>,
        otherId: Long,
        config: AutoLinkConfig
    ): CandidateEdge? {
        val shared = myTags.filter { it in otherTags }
        if (shared.size < config.tag.minOverlap) return null
        val union = (myTags + otherTags).toSet()
        val jaccard = shared.size.toDouble() / union.size
        return CandidateEdge(
            targetId = otherId,
            relation = AutoLinkRelations.SHARES_TAG,
            score = jaccard,
            metadata = mapOf("shared_tags" to shared, "overlap_count" to shared.size)
        )
    }

    private fun parseTags(raw: String?): List<String> =
        raw?.split(",")?.map { it.trim() }?.filter { it.isNotBlank() } ?: emptyList()
}

File: backend/mcp-code-intelligence-nodejs/src/memory/linking-strategies/tag-strategy.ts
/**
 * TagStrategy — shared tag overlap linking.
 * KSA-190: Auto-Linking Logic on KB Ingest.
 */

import Database from 'better-sqlite3';
import type { LinkingStrategy, CandidateEdge } from './types.js';
import type { AutoLinkConfig } from '../auto-link-config.js';

export class TagStrategy implements LinkingStrategy {
  readonly name = 'tag';
  private readonly db: Database.Database;

  constructor(db: Database.Database) {
    this.db = db;
  }

  isEnabled(config: AutoLinkConfig): boolean {
    return config.tag.enabled;
  }

  findCandidates(entryId: number, config: AutoLinkConfig): CandidateEdge[] {
    // Get tags for this entry
    const entry = this.db.prepare(
      'SELECT tags FROM knowledge_entries WHERE id = ?'
    ).get(entryId) as { tags: string } | undefined;
    if (!entry || !entry.tags) return [];

    const myTags = entry.tags.split(',').map(t => t.trim()).filter(Boolean);
    if (myTags.length < config.tag.minOverlap) return [];

    // Find entries with overlapping tags
    const conditions = myTags.map(() => 'tags LIKE ?').join(' OR ');
    const params = myTags.map(t => `%${t}%`);
    const rows = this.db.prepare(
      `SELECT id, tags FROM knowledge_entries WHERE id != ? AND archived = 0 AND (${conditions})`
    ).all(entryId, ...params) as Array<{ id: number; tags: string }>;

    const candidates: CandidateEdge[] = [];
    for (const row of rows) {
      const otherTags = row.tags.split(',').map(t => t.trim()).filter(Boolean);
      const shared = myTags.filter(t => otherTags.includes(t));
      if (shared.length >= config.tag.minOverlap) {
        const union = new Set([...myTags, ...otherTags]);
        const jaccard = shared.length / union.size;
        candidates.push({
          targetId: row.id,
          relation: 'SHARES_TAG',
          score: jaccard,
          metadata: { shared_tags: shared, overlap_count: shared.length },
        });
      }
    }

    candidates.sort((a, b) => b.score - a.score);
    return candidates.slice(0, config.tag.maxEdges);
  }
}
